In [1]:
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
from autodiff.backprop import *
from autodiff.forward import *
from autodiff.rootfinding import *

Backpropagation module is built upon the interfaces developed in central code file "Autodiff.forward".
It calculate the derivative of each nodes in the compuational graph with respect to the root nodes. Therefore with different root nodes, we should expect to see different values of derivative.
suppose we have the following structure:

 - $x = 1$, $y = 2$
 - $c = \sin(x)$
 - $d = c \cdot y$

Note: after one round of back propagation, the .bder attributes stores the answer from the last round until it is cleared when a new round is called upon.

In [2]:
x = Variable()
y = Variable()
c = sin(x)
d = c*y
back_propagation(c,{x:1,y:2})
print('derivative of x with respect to c is ', x.bder)
print('derivative of y with respect to c is ', y.bder)
back_propagation(d,{x:1,y:2})
print('derivative of x with respect to c is ', x.bder)
print('derivative of y with respect to c is ', y.bder)

derivative of x with respect to c is  0.5403023058681398
derivative of y with respect to c is  0
derivative of x with respect to c is  1.0806046117362795
derivative of y with respect to c is  0.8414709848078965


If we calculate by hand:

$
\begin{align}
\frac{dc}{dx} &= cos(1) = 0.54 \\
\frac{dc}{dy} &= 0 \\
\frac{dd}{dx} &= y*\frac{dc}{dx} = 2*cos(1) = 1.08\\
\frac{dd}{dy} &= c = sin(1) =  0.84 \\ 
\end{align}
$

Back propagation is also integrated with the function Newton's

In [7]:
result_d=newton_scalar(f,{x:1,y:-1},max_itr = 25,method = 'backward')

In [12]:
print('x:',result_D[x])
print('y:',result_D[y])
print('function value:',abs(f.evaluation_at({x:result_d[x],y:result_d[y]})))

x: 0.0
y: -1.6420926159343308
function value: 0.0
